File: 322-coin-change/322.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        dp = [math.inf] * (amount + 1)
        dp[0] = 0

        for i in range(1, amount + 1):
            minimal = math.inf
            for c in coins:
                if (i - c) >= 0:
                    minimal = min(minimal, dp[i-c])
            dp[i] = minimal + 1

        if dp[amount] == math.inf:
            return -1
        else:
            return dp[amount]

File: 322-coin-change/322.c
#define min(a,b) ((a) < (b) ? (a) : (b))

int coinChange(int* coins, int coinsSize, int amount){
    int f[amount + 1];

    for (int i = 0; i < amount + 1; ++i)
        f[i] = INT_MAX;

    f[0] = 0;

    for (int i = 1; i < amount + 1; ++i) {
        for (int j = 0; j < coinsSize; ++j) {
            if (coins[j] <= i)
                f[i] = min(f[i], f[i - coins[j]] + 1);
        }
    }

    if (f[amount] == INT_MAX)
        return -1;
    else
        return f[amount];
}
